### PR TITLE
Improve vmware_vm_shell module

### PR DESCRIPTION
**Pull request by flytreeleft**
##### SUMMARY
- Get and return the exit code and process information of the program which was executed through VMware tools.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
modules/cloud/vmware/vmware_vm_shell

##### ANSIBLE VERSION
```
[DEPRECATED] [defaults]hostfile: The key is misleading as it can also be a list of hosts, a directory or a list of paths. It will be removed in 2.8. As alternative use one of [inventory]        
ansible 2.4.0.0                                 
  config file = /etc/ansible/ansible.cfg        
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']                                                                                      
  ansible python module location = /usr/lib/python2.7/site-packages/ansible                      
  executable location = /usr/bin/ansible        
  python version = 2.7.13 (default, Jul 21 2017, 03:24:34) [GCC 7.1.1 20170630]
```

##### ADDITIONAL INFORMATION
With this improvement, the task can be failed when the program running in guest occurred some errors.

### CODE REVIEW

File: lib/ansible/modules/cloud/vmware/vmware_vm_shell.py [hunk 1]
```diff
@@ -195,10 +224,13 @@ def main():
         if not vm:
             module.fail_json(msg='VM not found')
 
-        msg = execute_command(content, vm, p['vm_username'], p['vm_password'],
-                              p['vm_shell'], p['vm_shell_args'], p['vm_shell_env'], p['vm_shell_cwd'])
+        (exitcode, msg) = execute_command(content, vm, p['vm_username'], p['vm_password'],
+                                            p['vm_shell'], p['vm_shell_args'], p['vm_shell_env'], p['vm_shell_cwd'])
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_vm_shell.py]: Overindented line.

File: lib/ansible/modules/cloud/vmware/vmware_vm_shell.py [hunk 2]
```diff
@@ -141,9 +144,35 @@ def execute_command(content, vm, vm_username, vm_password, program_path, args=""
...
-    return cmdpid
+    pm = content.guestOperationsManager.processManager
+    cmdpid = pm.StartProgramInGuest(vm=vm, auth=creds, spec=cmdspec)
+
+    if cmdpid > 0:
+        exitcode = pm.ListProcessesInGuest(vm=vm, auth=creds, pids=[cmdpid]).pop().exitCode
+
+        # If its not a numeric result code, it says None on submit
+        while (re.match('[^0-9]+', str(exitcode))):
+            time.sleep(5)
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_vm_shell.py:159]: Do we really want to wait for executed process ? I can imagine - if there is Windows Update or some other long running program, then playbook will run for that much amount of time, which I think is overkill.

**Review by Akasurde — CHANGES_REQUESTED** [review]

**Comment by flytreeleft** [comment]
@Akasurde Sorry, I have no idea how to fix the `indented` problem :-(. Just please tell me what to do for this.

File: lib/ansible/modules/cloud/vmware/vmware_vm_shell.py (hunk 2 above)
Comment (flytreeleft) [reply lib/ansible/modules/cloud/vmware/vmware_vm_shell.py:159]: Hum, I think both situations are reasonable. What about to add a new argument like `wait_for_shell_exit` to determine whether wait or not?

File: lib/ansible/modules/cloud/vmware/vmware_vm_shell.py [hunk 3]
```diff
@@ -195,10 +224,13 @@ def main():
         if not vm:
             module.fail_json(msg='VM not found')
 
-        msg = execute_command(content, vm, p['vm_username'], p['vm_password'],
-                              p['vm_shell'], p['vm_shell_args'], p['vm_shell_env'], p['vm_shell_cwd'])
+        (exitcode, msg) = execute_command(content, vm, p['vm_username'], p['vm_password'],
+                                        p['vm_shell'], p['vm_shell_args'], p['vm_shell_env'], p['vm_shell_cwd'])
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_vm_shell.py]: Something like this - 

```python
(exitcode, msg) = execute_command(content, vm, p['vm_username'], p['vm_password'],
                                  p['vm_shell'], p['vm_shell_args'], p['vm_shell_env'], p['vm_shell_cwd'])
```

**Pull request closed without merging**